### ISSUE TITLE
Inherit repo permissions from Gerrit

### ISSUE THREAD

**Issue by christinelovett** [sourcegraph/sourcegraph#23563]
#### Feature request description

Inherit repository permissions from Gerrit to Sourcegraph. 

#### Is your feature request related to a problem? If so, please describe.

Using the explicit permissions API for a single code host is not an optimal solution. It introduces operational overhead and relies on scripts that need to be maintained and validated.  Providing repository permissions on Sourcegraph repos that inherit directly from Gerrit would make for a better UX.

#### Describe alternatives you've considered.

Current workaround is using the explicit permissions API. 

#### Additional context

[Customer](https://docs.google.com/spreadsheets/d/1OEhzdMSlkGOaWyGKwdiAGlirsKKj9EN45Izn7kdKNTg/edit#gid=0&range=A1008)
Environment: 25 repos, 100 users
Permissions are currently: **Read only** at the repository level

**Comment by christinelovett** [comment]
How it works today:

- The groups of devs get access to Sourcegraph via the corp VPN
- In addition to this, they want to check repo permissions through gerrit 
- Gerrit only has 1 API to check 
- The permissions mapping is **user:repo**, so they have to make 20K calls everyday to set the permissions 
- Run a sync against the group on a daily basis, they will have changes every few days as users are added and removed
- There is a primary gerrit server, and a mirror server from which the code is pulled into Sourcegraph. The permissions have to be mapped from the primary gerrit server.

**Comment by dan-mckean** [comment]
Other known customers on Gerrit - [C](https://docs.google.com/spreadsheets/d/1OEhzdMSlkGOaWyGKwdiAGlirsKKj9EN45Izn7kdKNTg/edit#gid=0&range=A7103) and [L](https://docs.google.com/spreadsheets/d/1OEhzdMSlkGOaWyGKwdiAGlirsKKj9EN45Izn7kdKNTg/edit#gid=0&range=A7103)

**Comment by dan-mckean** [comment]
Repo management - the initial work here right now is to assess the viability of this, and the cost of doing the work. On that basis we'll prioritize it.

**Comment by dan-mckean** [comment]
Original summary of investigation posted to Slack [here](https://sourcegraph.slack.com/archives/C01NKL6V7E2/p1633603828013800) - but since that will only live for 90d, copying here.

---

Hi @Christine Lovett

Investigation revealed that supporting Gerrit permissions is bigger and nastier than hoped.

**TL;DR:**
There’s two approaches:
1. Only support repo-level permissions - up to 4 weeks for 2 engineers
2. Go down to the level of more granular permissions that Gerrit actually allows - up to 8 weeks for 2 engineers

There’s outstanding questions about the 3 known Gerrit-using customers here - the answers will inform whether we could stop at #1.

Even #1 is a big diversion from the plans (which include addressing enterprise scale and resilience, as well as several tactical things for the big strategic customers - like Perforce for Salesforce).

**Summary:**
It’s not something we can accommodate in what’s left of Q3. I think it might be something we can hope to start on in the 2nd half of Q4, though that will depend on progress against company objectives.

If it turns out that the other Gerrit-using customers are somehow blocked from expansion, or another prospect needs this, that might shift the balance.

**More info:**
What we did for Perforce to support more granular (than just repo level) permissions was a hack that results in us treating any directory that has more granular permissions (than the parent repo) as it’s own repo in Sourcegraph - that is because permissions are currently exclusively tied to repos in our system.

We’re already seeing pain (searches for example) as a result of that. So we’re not going that route again for Gerrit.

We’ve also got to define a new authz provider in order to sync the permissions, as well as all the interaction with Gerrit.
The teams lack of experience with authz providers as well as the Gerrit APIs is what makes them vary wary - even without the very unknown re-architecture work that would be needed to support permissions more granular than repo-level.

CC @Serina Clark

**Comment by unknwon** [comment]
_Just driving by_

>2. Go down to the level of more granular permissions that Gerrit actually allows - up to 8 weeks for 2 engineers

Assuming this approach is talking about branch and/or file-level permissions? If so, I'd say this could be a dangerous approach which may give customers wrong impression of our system is designed with such low-level permissions in mind (which we are not yet). Of course this does not mean we shouldn't explore solutions to support low-level permissions like branch or file.

**Comment by jplahn** [comment]
We are now tracking this work here: https://github.com/sourcegraph/sourcegraph/issues/32492#issuecomment-1064985094

I don't think we need to keep this issue open since we're aware of it and we're actively working on this now.
